### PR TITLE
Fix ASLR bug: update cached_mem_end after relocation and switch tee_svc_uref_base to use VCORE_START_VA

### PR DESCRIPTION
**Pull request by maximus64**
cached_mem_end was calculated before relocation. Add code to update cached_mem_end with ASLR load offset.

tee_svc_uref_base was using hardcoded TEE_TEXT_VA_START define value. This value no longer valid after TEE core relocation. Switch to use VCORE_START_VA which is linker variable that should get update after relocation code executed.

Signed-off-by: Khoa Hoang <admin@khoahoang.com>

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`
The first line of the commit could be updated to summarize the fix better, the `OPTEE` word also seems rather redundant.

**Comment by jforissier** [comment]
@jenswi-linaro +1
This patch does two things and would therefore better be split in two, think about what happens when one reads the output of git log --oneline, with the current subject you have no idea what is fixed, with two patches you could have:
"core: aslr: apply load offset to cached_mem_end"
"core: aslr: set tee_svc_uref_base to VCORE_START_VA"

`Acked-by: Jerome Forissier <jerome@forissier.org>`

**Comment by maximus64** [comment]
I updated the tags and split into two separate commit

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 1]
```diff
@@ -500,6 +500,13 @@ shadow_stack_access_ok:
 	 */
 	ldr	r0, =boot_mmu_config
 	ldr	r0, [r0, #CORE_MMU_CONFIG_LOAD_OFFSET]
+	/*
+	 * Update cached_mem_end address with load offset since it
+	 * calculated before relocation.
```
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_entry_a32.S]: minor: "... since it **was** calculated ..."

File: core/arch/arm/kernel/generic_entry_a64.S [hunk 2]
```diff
@@ -182,6 +182,14 @@ clear_nex_bss:
 	 * the memory will become write protected.
 	 */
 	ldr	x0, boot_mmu_config + CORE_MMU_CONFIG_LOAD_OFFSET
+	/*
+	 * Update cached_mem_end address with load offset since it
+	 * calculated before relocation.
```
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_entry_a64.S]: ditto

**Pull request merged**